Only truncate the hash to sign/verify by DSA when it is longer than the subgroup length

kailun-qin opened the pull request:
According to FIPS 186-3, section 4.6, the hash should be truncated to the
byte-length of the subgroup, crypto/dsa doesn't perform that truncation itself.

This patch proposes a fix to the current implementation, where the truncation is
always done to the byte-length of the subgroup for the hash to sign/verify, no
matter it has larger size or not. This may lead to runtime errors of `slice
bounds out of range`.

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

ProsaicSatsuma reviewed (approved):
Thanks @kailun-qin - good catch

kailun-qin committed:
Only truncate the hash to sign/verify by DSA when it is longer than the subgroup length

According to FIPS 186-3, section 4.6, the hash should be truncated to the
byte-length of the subgroup, crypto/dsa doesn't perform that truncation itself.

This patch proposes a fix to the current implementation, where the truncation is
always done to the byte-length of the subgroup for the hash to sign/verify, no
matter it has larger size or not. This may lead to runtime errors of `slice
bounds out of range`.

Signed-off-by: Kailun Qin <kailun.qin@intel.com>
